module/mailnews/mailread.c

```diff
@@ -353,3 +400,145 @@ void UserMailReply(int msg_num, Bool reply_all)
    MailSendReply(hReadMailDlg, reply);
 }
+
+void OnColumnClick(LPNMLISTVIEW pLVInfo)
+{
+   int nSortColumn = pLVInfo->iSubItem;
+   bool bSortAscending = TRUE;
+
+   if (nSortColumn == currentSortColumn)
+   {
+      // Toggle sort direction if the same column is clicked again
+      currentSortAscending = !currentSortAscending;
+   }
+   else
+   {
+      // Set ascending order for new column
+      currentSortColumn = nSortColumn;
+      currentSortAscending = TRUE;
+   }
+
+   LPARAM lParamSort = (currentSortAscending ? 1 : -1) * (currentSortColumn + 1);
+
+   if (hbmUpArrow != NULL && hbmDownArrow != NULL)
+   {
+      ListView_SetHeaderSortImage(pLVInfo->hdr.hwndFrom, currentSortColumn, currentSortAscending);
+   }
+
+   // Sort list
+   ListView_SortItemsEx(pLVInfo->hdr.hwndFrom, CompareMailListItems, lParamSort);
+}
+
+/****************************************************************************/
+/*
+ * CompareMailListItems: Comparison function for sorting items in the dialog
+ * list view.
+ */
+int CALLBACK CompareMailListItems(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSort)
+{
+   bool bSortAscending = (lParamSort > 0);
+   int nColumn = abs(lParamSort) - 1;
+   char szBuf1[MAX_SUBJECT], szBuf2[MAX_SUBJECT];
+   int time1 = 0;
+   int time2 = 0;
+
+   HWND hListView = GetDlgItem(hReadMailDlg, IDC_MAILLIST);
+
+   // Extract the msg_num values from the list view items
+   ListView_GetItemText(hListView, lParam1, COL_ORDER, szBuf1, sizeof(szBuf1));
+   ListView_GetItemText(hListView, lParam2, COL_ORDER, szBuf2, sizeof(szBuf2));
+   int msg_num1 = atoi(szBuf1);
+   int msg_num2 = atoi(szBuf2);
+
+   ListView_GetItemText(hListView, lParam1, nColumn, szBuf1, sizeof(szBuf1));
+   ListView_GetItemText(hListView, lParam2, nColumn, szBuf2, sizeof(szBuf2));
+
+   switch (nColumn)
+   {
+   case COL_ORDER:
+      return bSortAscending ? atoi(szBuf1) - atoi(szBuf2) : atoi(szBuf2) - atoi(szBuf1);
+   case COL_FROM:
+      return bSortAscending ? stricmp(szBuf1, szBuf2) : stricmp(szBuf2, szBuf1);
+   case COL_SUBJECT:
+      return bSortAscending ? stricmp(szBuf1, szBuf2) : stricmp(szBuf2, szBuf1);
+   case COL_TIME:
+      time1 = mail_date_map[msg_num1];
+      time2 = mail_date_map[msg_num2];
+
+      return bSortAscending ? (time1 - time2) : (time2 - time1);
+   default:
+      debug(("Unhandled column (column #%d) in CompareMailListItems\n", nColumn));
+      return 0;
+   }
+}
+
+/***************************************************************************/
+/*
+ * StringToTimestamp converts mail date strings to integers for comparison
+ */
+/***************************************************************************/
+int StringToTimestamp(const char *dateStr)
+{
+   struct tm tm = {0};
+
+   const char *localeName = setlocale(LC_ALL, "");
+
+   // Skip the day of the week
+   const char *remainingStr = dateStr + 4;
+
+   std::istringstream input(remainingStr);
+
+   input.imbue(std::locale(localeName));
```

Comment:
I've never seen some of these functions before.  Hopefully you can test this by forcing your locale to German and receiving some mail.

Reply:
Ah, I let the German client slip from my mind. Preliminary tests are all looking good (had to remove the NOSORT flag for the German resources) and I now know how to fire up the German client 👍🏻